src/Components/Web.JS/src/Services/WebRootComponentManager.ts

```diff
@@ -135,60 +140,104 @@ export class WebRootComponentManager implements DescriptorHandler, NavigationEnh
     }
 
     this._didWebAssemblyFailToLoadQuickly = true;
-    this.refreshAllRootComponents();
+    this.rootComponentsMayRequireRefresh();
   }
 
   private async startCircutIfNotStarted() {
-    if (this._hasStartedCircuit) {
+    if (hasStartedServer()) {
+      if (!isCircuitActive()) {
+        await startCircuit(this);
+      }
       return;
     }
 
-    this._hasStartedCircuit = true;
-    await startCircuit(this);
+    await startServer(this);
   }
 
   private async startWebAssemblyIfNotStarted() {
     this.startLoadingWebAssemblyIfNotStarted();
 
-    if (this._hasStartedWebAssembly) {
+    if (hasStartedWebAssembly()) {
       return;
     }
 
-    this._hasStartedWebAssembly = true;
     await startWebAssembly(this);
   }
 
-  private async refreshAllRootComponentsAfter(promise: Promise<void>) {
-    await promise;
-
-    if (!this._hasPendingRootComponentUpdate) {
-      this._hasPendingRootComponentUpdate = true;
-      setTimeout(() => {
-        this.refreshAllRootComponents();
-      }, 0);
-    }
-  }
-
   // This function should be called each time we think an SSR update
   // should be reflected in an interactive component renderer.
   // Examples include component descriptors updating, document content changing,
   // or an interactive renderer attaching for the first time.
-  private refreshAllRootComponents() {
-    this._hasPendingRootComponentUpdate = false;
-    this.refreshRootComponents(this._activeDescriptors);
+  private rootComponentsMayRequireRefresh() {
+    if (this._isComponentRefreshPending) {
+      return;
+    }
+
+    this._isComponentRefreshPending = true;
+    setTimeout(() => {
+      this._isComponentRefreshPending = false;
+      this.refreshRootComponents(this._rootComponents);
+    }, 0);
```

Comment:
This allows us to liberally call `rootComponentsMayRequireRefresh` without taking the small performance hit from frequent repeated calls to `refreshRootComponents`.

Reply:
That's good. Would be worth adding a comment for it.

The first time I saw this code I was worried it was some kind of hack (in the bad old days people used to do `setTimeout(..., 0)` as a way to fix order-of-execution issues despite it being impossible reason about) 😄